### ISSUE TITLE
Race condition when updating earnings / sale counts

### ISSUE THREAD

**Issue by pippinsplugins** [easydigitaldownloads/Easy-Digital-Downloads#2558]
It seems there are some discrepancies with the count stored in `_edd_download_sales` and the count retrieved from `EDD_Logging` for the `sale` log type.

For example, `_edd_download_sales` might report `155` but then the sales log table reports `169` sale records.

I've found a few items on the EDD site that have discrepancies and @jdevalk noticed it as well on a brand new product.

My guess is that the log count is incorrect, not the postmeta value. I'm not sure if the count is just wrong or there are actually extra records getting created.

Some counts might not be getting properly adjusted for refunds and deletions as well.

**Comment by pippinsplugins** [comment]
There's definitely too much potential for a race condition here (and for sales): 
```php

	/**
	 * Increase the earnings by the given amount
	 *
	 * @since 2.2
	 * @return float|false
	 */
	public function increase_earnings( $amount = 0 ) {
		$earnings = edd_get_download_earnings_stats( $this->ID );
		$earnings = $earnings + (float) $amount;
		if ( update_post_meta( $this->ID, '_edd_download_earnings', $earnings ) ) {
			$this->earnings = $earnings;
			return $earnings;
		}
		return false;
	}
```

**Comment by jdevalk** [comment]
Yeah this should simply be done with a direct MySQL query, abstracting it might be a good idea as even core has this problem.

**Comment by pippinsplugins** [comment]
Agreed.

**Comment by cklosowski** [comment]
:+1: for abstracting it in a direct query.

**Comment by pippinsplugins** [comment]
First pass submitted in #3078.

Any feedback?

**Comment by pippinsplugins** [comment]
I'm not sure why the unit tests are failing here . . . it works outside of unit tests.

**Comment by cklosowski** [comment]
Will have a look

**Comment by cklosowski** [comment]
First two errors look like rounding issues, 2nd two look like something isn't being updated.

**Comment by cklosowski** [comment]
@pippinsplugins  See commit and comment:
https://github.com/easydigitaldownloads/Easy-Digital-Downloads/commit/bd70164cf1141e3f005ccaab92e94a3624d9a4b1

**Comment by pippinsplugins** [comment]
:+1:

**Comment by pippinsplugins** [comment]
Merged into `release/2.3` for further testing.

**Comment by pippinsplugins** [comment]
I've tested this by using `wp edd payments create --id=6908 --status=complete --number=500` and all stats came out perfectly when it was finished. Closing!
